hedging/hedging.go

```diff
@@ -0,0 +1,126 @@
+package hedging
+
+import (
+	"errors"
+	"flag"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/cristalhq/hedgedhttp"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"golang.org/x/time/rate"
+)
+
+var (
+	ErrTooManyHedgeRequests       = errors.New("too many hedge requests")
+	totalHedgeRequests            prometheus.Counter
+	totalRateLimitedHedgeRequests prometheus.Counter
+	once                          sync.Once
+)
+
+// Config is the configuration for hedging requests.
+type Config struct {
+	// At is the duration after which a second request will be issued.
+	At time.Duration `yaml:"at"`
+	// UpTo is the maximum number of requests that will be issued.
+	UpTo int `yaml:"up_to"`
+	// The maximun of hedge requests allowed per second.
+	MaxPerSecond int `yaml:"max_per_second"`
+}
+
+// RegisterFlags registers flags.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("", f)
+}
+
+// RegisterFlagsWithPrefix registers flags with prefix.
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.IntVar(&cfg.UpTo, prefix+"hedge-requests-up-to", 2, "The maximun of hedge requests allowed.")
+	f.DurationVar(&cfg.At, prefix+"hedge-requests-at", 0, "If set to a non-zero value a second request will be issued at the provided duration. Default is 0 (disabled)")
+	f.IntVar(&cfg.MaxPerSecond, prefix+"hedge-max-per-second", 5, "The maximun of hedge requests allowed per seconds.")
+}
+
+// Client returns a hedged http client.
+// The client transport will be mutated to use the hedged roundtripper.
+func (cfg *Config) Client(client *http.Client) (*http.Client, error) {
+	return cfg.ClientWithRegisterer(client, prometheus.DefaultRegisterer)
+}
+
+// ClientWithRegisterer returns a hedged http client with instrumentation registered to the provided registerer.
+// The client transport will be mutated to use the hedged roundtripper.
+func (cfg *Config) ClientWithRegisterer(client *http.Client, reg prometheus.Registerer) (*http.Client, error) {
+	if reg == nil {
+		reg = prometheus.DefaultRegisterer
+	}
+	if client == nil {
+		client = http.DefaultClient
+	}
+	if cfg.At == 0 {
+		return client, nil
+	}
```

Comment:
personally, i'd move cases which result in a quick exit without doing anything to the top of the function because it makes it easy to create mental assertions about what value a variable can/can't have. But that might just be a personal habit